Nodejs/Product/Nodejs/Workspace/ContextMenuProvider.cs

```diff
@@ -0,0 +1,238 @@
+            {
+                // Unfortunately the NpmController (and NpmCommander), used for the install, update commands
+                // doesn't support running arbitrary scripts. And changing that is outside
+                // the scope of these changes.
+
+                var index = nCmdID - PkgCmdId.cmdidWorkSpaceNpmDynamicScript;
+                var packageJson = PackageJsonFactory.Create(((IFileNode)node).FullPath);
+
+                Debug.Assert(packageJson != null, "Failed to create package.json");
+
+                var scripts = packageJson.Scripts;
+                if (index < scripts.Length)
+                {
+                    var npmPath = NpmHelpers.GetPathToNpm();
+
+                    await NpmWorker.ExecuteNpmCommandAsync(
+                        npmPath,
+                        executionDirectory: node.Workspace.Location,
+                        arguments: new[] { "run-script", scripts[index].CommandName },
+                        visible: true); // show the CMD window
+                }
+            }
+
+            public bool QueryStatus(List<WorkspaceVisualNodeBase> selection, Guid pguidCmdGroup, uint nCmdID, ref uint cmdf, ref string customTitle)
+            {
+                if (selection.Count != 1 || !EnsurePackageJson(selection.First()))
+                {
+                    return false;
+                }
+
+                if (pguidCmdGroup == Guids.NodeToolsWorkspaceCmdSet)
+                {
+                    switch (nCmdID)
+                    {
+                        // we can always install missing, new, and update npm packages
+                        case PkgCmdId.cmdidWorkSpaceNpmInstallMissing:
+                        case PkgCmdId.cmdidWorkSpaceNpmInstallNew:
+                        case PkgCmdId.cmdidWorkSpaceNpmUpdate:
+                            cmdf = (uint)(OLECMDF.OLECMDF_SUPPORTED | OLECMDF.OLECMDF_ENABLED);
+                            return true;
+                    }
+
+                    // Each id we return true for, this method is called with id + 1
+                    // this way we can add each script to the context menu
+                    if (nCmdID >= PkgCmdId.cmdidWorkSpaceNpmDynamicScript && nCmdID < PkgCmdId.cmdidWorkSpaceNpmDynamicScriptMax)
+                    {
+                        var node = (IFileNode)selection.First();
+
+                        if (QueryDynamic(nCmdID, node.FullPath, ref customTitle))
+                        {
+                            cmdf = (uint)(OLECMDF.OLECMDF_SUPPORTED | OLECMDF.OLECMDF_ENABLED);
+                            return true;
+                        }
+                    }
+                }
+
+                return false;
+            }
+
+            private bool QueryDynamic(uint nCmdID, string filePath, ref string customTitle)
+            {
+                var index = nCmdID - PkgCmdId.cmdidWorkSpaceNpmDynamicScript;
+                var packageJson = PackageJsonFactory.Create(filePath);
```

Comment:
Does this end up re-reading the package.json file for every script?

Reply:
Also, does it re-read every time you right click? Or is there a change it could show stale scripts when you right-click if the package.json file got modified?

Reply:
Yes, we do... looks like there isn't a better way, without running the risk of stale commands. (And considering a package.json is not huge, this is probably fine).